mirascope/core/base/toolkit.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+from abc import ABC
+from textwrap import dedent
+from typing import Callable, Optional, ClassVar, ParamSpec
+
+from pydantic import BaseModel, ConfigDict
+from typing_extensions import LiteralString
+
+from . import BaseTool
+from .._internal.utils import get_template_variables, convert_function_to_base_tool
+
+_TOOLKIT_TOOL_METHOD_MARKER: LiteralString = "__toolkit_tool_method__"
+
+P = ParamSpec("P")
+
+
+def toolkit_tool(
+        method: Callable[[BaseToolKit, ...], str],
+) -> Callable[[BaseToolKit, ...], str]:
+    # Mark the method as a toolkit tool
+    setattr(method, _TOOLKIT_TOOL_METHOD_MARKER, True)
+
+    return method
+
+
+class BaseToolKit(BaseModel, ABC):
+    """A class for defining tools for LLM call tools."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    _toolkit_tool_method: ClassVar[Callable[[BaseToolKit, ...], str]]
+    _toolkit_template_vars: ClassVar[list[str]]
+
+    def create_tool(self) -> type[BaseTool]:
+        """The method to create the tools."""
+        formated_template = self._toolkit_tool_method.__doc__.format(self=self)
+        return convert_function_to_base_tool(self._toolkit_tool_method, BaseTool, formated_template)
+
+    @classmethod
+    def __pydantic_init_subclass__(cls, **kwargs):
+        toolkit_tool_method: Optional[Callable] = None
+        for name, value in cls.__dict__.items():
+            if getattr(value, _TOOLKIT_TOOL_METHOD_MARKER, False):
+                if toolkit_tool_method:
+                    raise ValueError("Only one toolkit_tool method is allowed")
+                toolkit_tool_method = value
+        if not toolkit_tool_method:
+            raise ValueError("No toolkit_tool method found")
+
+        # Validate the toolkit_tool_method
+        if (template := toolkit_tool_method.__doc__) is None:
+            raise ValueError("The toolkit_tool method must have a docstring")
+
+        dedented_template = dedent(template).strip()
+        if not (template_vars := get_template_variables(dedented_template)):
+            raise ValueError("The toolkit_tool method must have template variables")
+
+        if dedented_template != template:
+            toolkit_tool_method.__doc__ = dedented_template
+        for var in template_vars:
+            if not var.startswith("self."):
+                # Should be supported un-self variables?
+                raise ValueError(
+                    "The toolkit_tool method must use self. prefix in template variables"
+                )
+
+            self_var = var[5:]
+            # Expecting pydantic model fields or class attribute and property
+            if self_var in cls.model_fields or hasattr(cls, self_var):
+                continue
+            raise ValueError(
+                f"The toolkit_tool method template variable {var} is not found in the class"
+            )
+
+        cls._toolkit_tool_method = toolkit_tool_method
```

Comment:
before we set the methods, we should also make sure that if provided the user can create a namespace where the tools name becomes `{namespace}.{fn_name}`

Reply:
Which of these cases are of concern?
My assumption is that the user has already defined a method and just takes it out and stores it in the tool's call method, but I didn't see the need to check the name.

Reply:
The user should be able to specify a namespace for a toolkit that updates the name of every function inside to contain the namespace.

For example:

```python
class FileSystem(Toolkit):
    namespace = "fs"

    @toolkit_tool
    def mkdir(self, ...):
        ...

print(FileSystem().create_tools()[0].__name__)
#> fs.mkdir
```

By default, `namespace` should be `None`, but if set, we should update the function names to be under the specified namespace. Does this make sense?

Reply:
I see.
As for the namespace declaration, is it OK to use namespace: ClassVar[str] as in the example?

>  we should also make sure that if provided the user can create a namespace where the tools name becomes {namespace}.{fn_name}

I'm a little confusing because the `__name__` of a function or method usually does not contain a namespace.

`assert inspect.signature.__name__ == 'signature'`

Reply:
Correct, we are doing this for the purposes of creating a "tool" for the LLM.

Consider two toolkits:

```python
class Library(Toolkit):
    books: list[Book]

    def find(self, title: str) -> Book:
        ...

class FileSystem(Toolkit):
    files: list[File]

    def find(self, filename: str) -> File:
        ...
```

We want to be able to insert both of the `find` methods, but we need to be able to identify which method is which, and we also should differentiate so the LLM can also better understand its tools. If we namespace, the LLM will see tools like `lib.find` and `fs.find`.

Reply:
Perhaps rather than updating the name of the function with the namespace we should instead pass the namespace into the `convert_function_to_base_tool` method like you did with `__doc__` and update the `_name` classmethod to use the namespaced name, that way there won't be a weird mismatch with the Python defaults / expectations.

Reply:
I see. I understand the situation.

So what scope is affected by namespace?
In other words, where should I manage the name of this namespace?
Is it the whole module? Or the tool_kit decorator?

I don't see a complete use case, but it seems fine to keep it at module level.

```python
__namespace: set[str] = set()
...

print(__namespace)
#   {'lib.find', 'fs.find'}
```

Reply:
I will add new comments directly where they are applicable